src/model.js

```diff
@@ -0,0 +1,187 @@
+/*
+  model.js
+  This file is required. It must export a class with at least one public function called `getData`
+*/
+const request = require('request-promise').defaults({ gzip: true, json: true });
+const { formatFeature, serializeQueryParams, convertToWGS84 } = require('./helpers/utils');
+const ArcgisSearchProviderError = require('./helpers/arcgis-search-provider-error');
+const logger = require('./helpers/logger');
+const { objTypeGeometryQuerySchema, arrayTypeGeometryQuerySchema } = require('./helpers/geometry-schema');
+
+const MAX_PAGE_SIZE = 100; // maximum number of results returned from portal per request
+const FIELD_DICTIONARY = require('./helpers/field-dictionary');
+const PORTAL_URL = 'http://www.arcgis.com/sharing/rest/search';
+const PROJ4_WKIDS = [4326, 4269, 3857, 3785, 900913, 102113]; // supported wkids by proj4
+
+class ArcgisSearchModel {
+  // Main getData method which is used to send data to Koop 
+  async getData(req, callback) {
+    try {
+      const portalQuery = this.buildPortalQuery(req.query);
+      const items = await this.getPortalItems(PORTAL_URL, portalQuery, MAX_PAGE_SIZE);
+      const geojson = this.getGeoJson(items, FIELD_DICTIONARY);
+      callback(null, geojson);
+    } catch (error) {
+      throw new ArcgisSearchProviderError(
+        error?.message || 'Error in Arcgis Search Provider',
+        error?.statusCode || 500
+      );
+    }
+  }
+
+  buildPortalQuery(requestQuery) {
+    const portalQuery = { f: 'json' };
+    // TODO: ensure appropriate quotes for Search in Online
+    // Remove any variant of 1=1 as it is not recognized by AGO search; replace = with; replace and ' with "
+    portalQuery.q =
+      (requestQuery.where || '*')
+        .replace(/1=1|(\(1=1\))|(AND\s1=1)|(AND\s\(1=1\))/g, '')
+        .replace(/\s+=\s+/g, ':')
+        .replace(/'/g, '"')
+        .replace(/[ \t]+$/, '');
+
+    portalQuery.num = requestQuery.resultRecordCount || 100;
+    portalQuery.start = requestQuery.resultOffset || 1;
+
+    if (requestQuery.orderByFields) {
```

Comment:
anything coming in on the query object should be validated in a separate distinct step at the beginning of getData.  That way you can have one place where you throw 400 errors or log. So you'll want to have a validation for any query parameters you want to support.

Reply:
I have added a validation function but here, we aren't really validating. We are building portal query if certain fields in the request query exist. For instance, if there is `orderByfields` in request query then, we add `sortOptions` to portal query. Not necessary validating or throwing error as absence of such fields does not necessarily mean the request is invalid and we would still want to full fill such requests or pass default values.

Reply:
As an example, let's say some submits an `orderByFields={foo:'bar'}`.  Don't we want to message that it isn't a valid value for `orderByFields`?

Reply:
That is true. I have added validation for `orderByFields`. Got valid sort fields from the portal page https://developers.arcgis.com/rest/users-groups-and-items/search.htm